### ISSUE TITLE
Overriding the listValues() method

### ISSUE THREAD

**Issue by mattsimpson** [ifsnop/mysqldump-php#101]
Hello Diego,

First off, thank you for the great code here. It's been extremely useful thus far, and I'm hoping to utilize it in [Entrada](http://entrada.org) for a developer tool I'm writing to create new installer SQL files.

I've run into an issue though that I'm trying to solve. There are several rows in several tables where the installer uses MySQL functions. For example, in the original `install.sql` file the query is:

``` sql
INSERT INTO `system_roles` VALUES (1,(YEAR(CURRENT_DATE())+4),1);
```

Of course once I do the dump of the data and replace `installer.sql`, that becomes:

``` sql
INSERT INTO `system_roles` VALUES (1,'2020'),1);
```

I was hoping there was a way to override the `listValues()` method so I could put in certain `$value_exceptions` where it would use the provided value instead of the value from the actual field.

The problem is, they are all `private` methods and variables in ifsnop/mysqldump-php.

Can you think of any other way? I'm happy to write the code and submit a pull request to you if you that would be useful?

**Comment by ifsnop** [comment]
Hi Matt,

I'm not sure I get your point. Standard mysqldump doesn't have a similar
option either. Please go ahead and show some code. :D

**Comment by bomoko** [comment]
@ifsnop I was going to open up an issue to something very closely related to this - but thought I'd check here first.

First, thank you for what is a very useful piece of software.

Now on to the question. Essentially, we're looking at using mysqldump-php to write a customized GDPR compliant DB export tool. What I'd like to do is to be able to simply override some of the methods that are presently marked private - such as listValues().

First possible solution would be marking listValues protected rather than private so that we could override it in a derived class.

A second possible solution would be to add a "transform" function that does nothing in your class, but which (being protected) might be used to override values in classes derived from Mysqldump.
Something like #138 (this is just a demonstration)

**Comment by ifsnop** [comment]
I think I prefer the second option. Adding some books to existing code
looks cleaner and useful. Do you need to modify other values? For example,
column names?

**Comment by bomoko** [comment]
Hi @ifsnop - I agree, hooks are far cleaner.
For our purposes, no - it's really just the field values themselves we need to be able to override.

**Comment by mattsimpson** [comment]
Hi all, I don't think I've done the nicest possible solution to this, but what I do have in here:

https://github.com/EntradaPackages/mysqldump-php/tree/feature/overriding-listValues-101

You're welcome to check it out.

**Comment by bomoko** [comment]
Hey @mattsimpson, thanks (and thanks for opening this issue initially)!

I think the virtue of adding a hook (rather than changing mysqldump-php _itself_) is that it can allow overriding in a variety of different ways. For instance, what we're experimenting with is extending the Mysqldump class and integrating the Faker library - I've written up a proof of concept and it works beautifully. Your extension could equally be implemented really easily.

Ideally we don't want to affect this projects mission to achieve parity with mysqldump itself. But being able to extend it would be amazing.

**Comment by mattsimpson** [comment]
That's great to hear @bomoko. Thanks!

**Comment by ifsnop** [comment]
If you can, make it generic, so it can be used in other places (not just to
modify the values of columns). When can iterate from there, since it is not
going to brake functionality. Maybe a hook prefix in the function name
could help?

Thanks both!

--
-- Madrid / Spain

**Comment by bomoko** [comment]
Brilliant - thanks to you both too.
I'll do some work and submit a new PR sometime during the next 24 hours or so.

**Comment by bomoko** [comment]
Okay, I've submitted a new PR that I'm slightly more happy with.

It's not as generic as you might like, yet, Diego, but I'm happy to work on this (and, indeed, on building an entire hook API, if you're keen!).

**Comment by bomoko** [comment]
Implemented in #139
